<a href="https://colab.research.google.com/github/ethangoldbeck/grit-hub/blob/master/Ethan_Goldbeck_HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error

In [0]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
alldata = pd.read_csv('drive/My Drive/Colab Notebooks/finalmaster-ratios.csv')

In [0]:
alldata.head()

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397
3           94     251724  505.585483  ...  43.731067  38.851729  40.427349
4            0      37382  495.586111  ...  36.118530  31.603714  19.648989

[5 rows x 190 columns]

In [0]:
allvariablenames = list(alldata.columns.values)

In [0]:
predictors = alldata[allvariablenames[8:len(allvariablenames)]]

In [0]:
target = alldata['# Purchases']   

In [0]:
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)   

In [0]:
model = LassoLarsCV(cv=10, precompute = False)
lasso = model.fit(predictors,target)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=8.526e-01, with an active set of 8 regressors, and the smallest cholesky pivot element being 8.816e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 12 iterations, i.e. alpha=4.263e-01, with an active set of 12 regressors, and the smallest cholesky pivot element being 8.816e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 12 iterations, i.e. alpha=4.263e-01, with an active set of 12 regressors, and the smallest cholesky pivot element

In [0]:
predictors_model=pd.DataFrame(allvariablenames[8:len(allvariablenames)])
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.5330720656233966]
['B01001036' 1.6591227573088478]
['B01001037' 2.4482060101637564]
['B01001038' 2.21381020716577]
['B02001005' 0.9859155344708008]
['B02001007' 0.006591090821753925]
['B13014016' 0.0061795097515955225]
['B13014026' 0.574990933842321]
['B13014027' 1.2849909864550721]
['B13016008' 0.31692716289106393]
['B15002027' 3.9418447021477347]
['B19001017' 1.8860806124452758]


Q1. Line by line the above code does the following:

1 - Create a pandas dataframe with 1 column of all the predictor variables

2 - rename the one column in the pd dataframe 'label'

3 - create a new column in the pd dataframe with all our model's coefficiants and name that column coeff (now we have a dataframe with two columns our predictor labels and their matching coefficiants in our model)

4 - Iterate through our pd dataframe and print the label and coeff for all labels that having a coefficiant greater than 0 (meaning they have a positive impact on our sales)

Q2. Variables/Coefficiants (most predictive at top):

B15002027 - 3.9418447021477347 - Females who went to school through 12th Grade but did not graduate

B01001037 - 2.4482060101637564 - Females aged 35 to 39

B01001038 - 2.21381020716577 - Females aged 40-44

B19001017 - 1.8860806124452758 - Household income of $200,000 or more (last 12 months)

B01001036 - 1.6591227573088478 - Females aged 30-34

B13014027 - 1.2849909864550721 - Woman 15 - 50 who did not have birth in the last 12 months who are unmarried and have a graduate or professional degree

B02001005 - 0.9859155344708008 - Asian Alone

B13014026 - 0.574990933842321 - 	Women 15 to 50 Years Who Had a Birth in the Past 12 Months with a bachelors degree

B01001014 - 0.5330720656233966 - Men aged 40 to 44

B13016008 - 0.31692716289106393 - Women 40-44 who had birth in the last 12 months

Starting with the first variable being the most impactful on sales, the above populations being present would lead to an increase in sales.

Q3. If I had to highlight two variable they would be 

B15002027 - 3.9418447021477347 - Females who went to school through 12th Grade but did not graduate

and

B01001037 - 2.4482060101637564 - Females aged 35 to 39

However, since the next most predictive varible is females aged 40-44 and it is almost as preductive as females aged 35-39, in a real life situation I would probably highlight these two variables together and choose  

Females who went to school through 12th Grade but did not graduate

and 

Females aged 35 to 44

In [0]:
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
21060.761034858362


In [0]:
train_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('test data MSE')
print(train_error)

test data MSE
36561.48369809522


Q4. Our test MSE is almost twice as large as our training MSE.  Considering the MSEs and the fact that the test data is less than half the size of our training data I would say our current model is overfitted for our training data and therefore would be a less stable test on other new data

In [0]:
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2733096196457573


Q5. If asked how predictive census data is on sales I would answer:

Census data is slightly predictive of our sales.  Using census data we can accurately predict about 25% of the variance in our sales.  This means it could be a good base for future prediction - but definitely can't be our predictive model on its own.

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
-3.9606645084610435


Q6. This Y intercept means that when no one from the census is present we are selling -3.96 units.  Obviously this is not possible. Practically this means that once we start adding populations to our model, in order to see actual sales we will need to go through a threshold of roughly 4 units. 